### PR TITLE
[Gutenberg Data Tutorial] Part 3: Building an edit form

### PR DESCRIPTION
**Pull request by adamziel**
This PR introduces the next part of [Create your First App with Gutenberg Data](https://github.com/WordPress/gutenberg/pull/38250) tutorial – the focus is on editing data.

[Rich-text preview](https://github.com/WordPress/gutenberg/blob/docs/data-basics-3-building-an-edit-form/docs/how-to-guides/data-basics/3-building-an-edit-form.md)

All feedback is very welcome!

cc @getdave @bph @ironprogrammer @ryanwelcher @mkaz @youknowriad @gziolo @talldan @tellthemachines @draganescu @gziolo @youknowriad @talldan @annezazu @noisysocks

### CODE REVIEW

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 1]
```diff
@@ -0,0 +1,500 @@
+# Building an edit form
+
+This part is about adding an *Edit* feature to our app. Here's a glimpse of what we're going to build:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md:3]: Do you want to link back to previous parts in case someone comes here first by accident?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+			<thead>
+				<tr>
+					<td>Title</td>
+					<td style={{width: 120}}>Actions</td>
+				</tr>
+			</thead>
+			<tbody>
+				{ pages?.map( ( page ) => (
+					<tr key={page.id}>
+						<td>{ page.title.rendered }</td>
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]: Do we need to `decodeEntities` here or does `.rendered` handle that?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+				<Button onClick={ onCancel } variant="tertiary">
+					Cancel
+				</Button>
+			</div>
+		</div>
+	);
+}
+```
+
+Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two using a component that requires the least code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two together using a component that requires the least code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:
```

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+### Step 3: Populate the form with page details
+
+We want the `EditPageForm` to display the title of the currently edited page. You may have noticed that it doesn't receive a `page` prop, only `pageId`. That's okay. Gutenberg Data allows us to easily access entity records from any component.
+
+In this case, we need to use the [`getEntityRecord`](/docs/reference-guides/data/data-core/#getentityrecord) selector. The list of records is already available thanks to the `getEntityRecords` call in `MyFirstApp`, so there won't even be any additional HTTP requests involved – we'll get the cached record right away.
+
+Here's how you can try it in your browser's dev tools:
+
+```js
+wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', 9 );  // Replace 9 with an actual page ID
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md:128]: Love these kinds of examples 👏

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+Now it should look like that:
+
+![](./media/edit-form/form-populated.png)
+
+### Step 4: Making the Page title field editable
+
+There's one problem with our _Page title_ field: you can't edit it. It receives a fixed `value` but doesn't update it when typing. We need an `onChange` handler.
+
+
+You may have seen a pattern similar to this one in other React apps:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
You may have seen a pattern similar to this one in other React apps. It's known as a ["controlled component"](https://reactjs.org/docs/forms.html#controlled-components):
```

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+	return (
+		<TextControl
+			value={ title }
+			onChange={ setTitle }
+		/>
+	);
+}
+```
+
+Updating entity records in Gutenberg Data is similar. We don't use `useState`, though. The counterpart of `setTitle` is an `editEntityRecord` action which stores the updates in the Redux state. Here's how you can try it out in your browser's dev tools:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
Updating entity records in Gutenberg Data is similar but instead of using `setTitle` to store in local state, we use the `editEntityRecord` action which stores the updates in the _Redux_ state. Here's how you can try it out in your browser's dev tools:
```

I think we can simplify the wording and shorten the sentance.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+
+```js
+// Get a real page ID first
+const pageId = wp.data.select( 'core' ).getEntityRecords( 'postType', 'page' )[0].id;
+
+// Uupdate the title
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
// Update the title
```

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+		/>
+	);
+}
+```
+
+Updating entity records in Gutenberg Data is similar. We don't use `useState`, though. The counterpart of `setTitle` is an `editEntityRecord` action which stores the updates in the Redux state. Here's how you can try it out in your browser's dev tools:
+
+```js
+// Get a real page ID first
+const pageId = wp.data.select( 'core' ).getEntityRecords( 'postType', 'page' )[0].id;
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md:178]: How come you used `getEntityRecords` here and not `getEntityRecord`? Does the later not expose the ID?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+
+// Uupdate the title
+wp.data.dispatch( 'core' ).editEntityRecord( 'postType', 'page', pageId, { title: 'updated title' } );
+```
+
+At this point, you may ask _How is `editEntityRecord` this better than `useState`?_ It offers a few features you wouldn't otherwise get. We can save the changes as easily as we retrieve the data and ensure that all caches will be correctly updated. In addition, the changes applied via `editEntityRecord` are easily undo-able via the `undo` and `redo` actions. Finally,  Because the changes live in the Redux state, other components can access them. For example, we could make the `PagesList` display the currently edited title.
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
At this point, you may ask _how is `editEntityRecord` better than `useState`? The answer is that it offers a few features you wouldn't otherwise get. 

Firstly, we can save the changes as easily as we retrieve the data and ensure that all caches will be correctly updated. 

Secondly, the changes applied via `editEntityRecord` are easily undo-able via the `undo` and `redo` actions. 

Lastly, because the changes live in the _Redux_ state, they are "global" and can be accessed by other components. For example, we could make the `PagesList` display the currently edited title.
```

Tweaked this. If you reject the formatting changes (no problem!) please note the grammar fixes towards the beginning.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+
+To that last point, let's see what happens when we use `getEntityRecord` to access the entity record we just updated:
+
+```js
+wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', pageId ).title
+```
+
+It doesn't reflect the edits. What's going on?
+
+Let's imagine it would reflect the edits. Then, anything you typed in the title field would be immediately displayed inside the `PagesList`. Gutenberg Data solves this problem by distinguishing between *Entity Records* and *Edited Entity Records*. Calling `getEntityRecord` always returns the data retrieved from the API. To see the edits applied, we need to use `getEditedEntityRecord` instead:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
Well, let's imagine for a second that it would reflect the edits. If it did, then anything you typed in the title field would be _immediately_ displayed inside the `PagesList`. 

Gutenberg Data solves this problem by distinguishing between *Entity Records* and *Edited Entity Records*. 

Calling `getEntityRecord` always returns the data retrieved _from the API_. To see the edits applied, we need to use `getEditedEntityRecord` instead:
```

Comment (getdave) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: Also I'm still not clear here as to the difference between the two. I was confused by the phrase "To see the edits applied".

Can we try to improve on the distinction between the two types of data? Is one store only in Redux state and then sync up to the API at a certain point? Is the other just whatever is on the API and ignores the Redux state?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 11]
```diff
@@ -0,0 +1,500 @@
...
+```
+
+Let's update `EditPageForm` accordingly. We can access the actions using the [`useDispatch`](/packages/data/README.md#usedispatch) hook similarly to how we use `useSelect` to access selectors:
+
+```js
+import { useDispatch } from '@wordpress/data';
+
+export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
+	const page = useSelect(
+		select => select( coreDataStore ).getEditedEntityRecord( 'postType', 'page', pageId ),
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md:227]: Right so this value comes from the Redux state so it's always "up to date" when you're editing.

If you'd run `getEntityRecord` it would show you what's currently on the server.

Question: when are the two synced up or is that a race condition kind of thing?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 12]
```diff
@@ -0,0 +1,500 @@
...
+
+It works, but there's still one thing to fix: the form modal doesn't automatically close because we never call `onSaveFinished`. Lucky for us, `saveEditedEntityRecord` returns a promise that resolves once the save operation is finished. Let's take advantage of it in `EditPageForm`:
+
+```js
+export function EditPageForm( { pageId, onCancel, onSaveFinished } ) {
+	// ...
+	const handleSave = async () => {
+		await saveEditedEntityRecord( 'postType', 'page', pageId );
+		onSaveFinished();
+	};
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md:311]: Nit: This is an interesting pattern. I've recently seen a problem with this whereby if the Modal is manually closed whilst the request is still running then you see an error.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 13]
```diff
@@ -0,0 +1,500 @@
...
+	const handleChange = ( title ) => editEntityRecord( 'postType', 'page', pageId, { title, date: -1 } );
+	// ...
+}
+```
+
+Once you refresh the page, open the form, change the title, and hit save, you should see the following error message:
+
+![](./media/edit-form/form-error.png)
+
+Fantastic! We can now restore the previous version of `handleChange` and move on to the next step.
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
Fantastic! We can now **restore the previous version of `handleChange`** and move on to the next step.
```

**Review by getdave — COMMENTED** [review]
🥇 This is great stuff Adam. It's easy to follow and seems to cover off all the common pain points.

I learnt quite a few things myself!

Some nits and suggestions left.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 12 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md:311]: Good point! I'd cover that in the upcoming more advanced tutorial and for now just assume noone will close it :-) I don't want to muddy the picture too much.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 8 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md:178]: `getEntityRecord` does expose the ID, but it also needs one as an argument. What do I pass there? I used `getEntityRecords` to find the first available ID – it seems like this snippet could use a better explanation.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 2 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: You're right! I just added that.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 8 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md:178]: Does it read better now @getdave ?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 10 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: I made a clearer distinction between the two – does it make more sense now?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 11 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md:227]: I added the following line:

Entity records are updated to reflect any saved changes right after the REST API request is finished.

Is it clear, or do you think we should talk a bit more about that?

**Comment by adamziel** [comment]
I rebased and addressed the feedback. Would you mind re-reviewing @getdave? 🙏

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 11 above)
Comment (getdave) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md:227]: Sounds good 👍

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 10 above)
Comment (getdave) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: That's better thanks.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 14]
```diff
@@ -0,0 +1,521 @@
+# Building an edit form
+
+This part is about adding an *Edit* feature to our app. Here's a glimpse of what we're going to build:
+
+![](./media/edit-form/form-finished.png)
+
+### Step 1: Add an _Edit_ button
+
+We can't have an *Edit* form without an *Edit* Button, so let's start by adding one to our `PagesList` component:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
We can't have an *Edit* form without an *Edit* button, so let's start by adding one to our `PagesList` component:
```

Nit: lowercase unless we're refering to the component in which case I'd recommend `<Button>`.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 15]
```diff
@@ -0,0 +1,521 @@
...
+				<Button onClick={ onCancel } variant="tertiary">
+					Cancel
+				</Button>
+			</div>
+		</div>
+	);
+}
+```
+
+Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two together using a component that requires the least code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
Now let's make the button display the form we just created. As this tutorial is not focused on web design, we will wire the two together using a component that requires the least amount of code: [`Modal`](https://developer.wordpress.org/block-editor/reference-guides/components/modal/). Let's update `PageEditButton` accordingly:
```

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+	return (
+		<TextControl
+			value={ title }
+			onChange={ setTitle }
+		/>
+	);
+}
+```
+
+Updating entity records in Gutenberg Data is similar but instead of using `setTitle` to store in local state, we use the `editEntityRecord` action which stores the updates in the _Redux_ state. Here's how you can try it out in your browser's dev tools:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]: Just thinking...will people know what Redux is and how it's related to WP data?

Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
Updating entity records in Gutenberg Data is similar but instead of using `setTitle` to store in local (component level) state, we use the `editEntityRecord` action which stores the updates in the _Redux_ state. Here's how you can try it out in your browser's dev tools:
```

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+
+To that last point, let's see what happens when we use `getEntityRecord` to access the entity record we just updated:
+
+```js
+wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', pageId ).title
+```
+
+It doesn't reflect the edits. What's going on?
+
+Well, let's imagine for a second that it would reflect the edits. If it did, then anything you typed in the title field would be _immediately_ displayed inside the `PagesList`.
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]: I think we need to add some wording here to clarify that the changes you type into the form field would be reflected in the PageList before you were ready to "commit" or persist them.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+
+```js
+wp.data.select( 'core' ).getEditedEntityRecord( 'postType', 'page', pageId ).title
+// "updated title"
+
+wp.data.select( 'core' ).getEntityRecord( 'postType', 'page', pageId ).title
+// { "rendered": "<original, unchanged title>", "raw": "..." }
+```
+
+As you can see, the `title` of an Entity Record is an object, but the `title` of an Edited Entity record is a string.
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md:214]: Why do we have inconsistent response shapes? Seems like a recipe for bugs...

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+
+This is no accident. Fields like `title`, `excerpt`, and `content` can only be rendered on the server so the REST API exposes both the `raw` markup and the `rendered` string. For example, in the block editor, `content.rendered` could used as a visual preview, and `content.raw` could be used to populate the code editor.
```
Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]: > can only be rendered on the server

How so? I wasn't clear here.

Comment (getdave) [inline docs/how-to-guides/data-basics/3-building-an-edit-form.md]:
```suggestion
This is no accident. Fields like `title`, `excerpt`, and `content` can only be rendered on the server and so the REST API exposes both the `raw` markup _and_ the `rendered` string. For example, in the block editor, `content.rendered` could used as a visual preview, and `content.raw` could be used to populate the code editor.
```

**Review by getdave — APPROVED** [review]
This reads really well.

I've left some minor suggestions and nits but otherwise I'd say this is good to go 🚢

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 16 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: Good question! I don't intend to explain that here, but a link or two could help 👍

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 19 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: Thanks for noting! I'll revise this using a few lines from the [TypeScript types PR](https://github.com/WordPress/gutenberg/pull/38666/) where this is explained more clearly.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 18 above)
Comment (getdave) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md:214]: Obviously not something we can/will fix in this PR.

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 16 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: I'm not sure how I feel about linking each mention of Redux, so I went for adding the following line to step 1:

> Also, this tutorial will lean heavily on Redux concepts such as state, actions, and selectors. If you are not familiar with them, you may want to start by reviewing [Getting Started With Redux](https://redux.js.org/introduction/getting-started).

Do you think that is sufficient?

File: docs/how-to-guides/data-basics/3-building-an-edit-form.md (hunk 17 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/3-building-an-edit-form.md]: I made some tweaks – is it clearer now?

**Comment by adamziel** [comment]
@getdave I'll go ahead and ship this. I'd still love to hear your feedback – I'll address any notes you may have in a follow-up PR.

**Comment by getdave** [comment]
Thanks for addressing all the feedback. Another great resource shipped. 

👏👏👏👏👏👏👏👏👏👏👏👏👏👏
